config: define config scopes in configuration

tgamblin opened the pull request:
This reworks the hard-coded configuration scopes in Spack to be defined in configuration. This allows scope hierarchy to be defined in one place and overridden in any config scope with sufficient precedence.


## Config scopes as config

All configuration now stems from `$spack/etc/spack` and `$spack/etc/spack/defaults`.
* Scopes included in configuration can be named, which allows the builtin site, user, system, etc. scopes to be configuration rather than hard-coded.
* `$spack/etc/spack/defaults` is the lowest priority
*  `$spack/etc/spack` includes other scopes, like `~/.spack`, at lower precedence than itself.
* You can override any scopes but default with `include::`, e.g. `include::[]` in an environment allows you to ignore everything but defaults entirely.

Here is `$spack/etc/spack/include.yaml`:

```yaml
include:
  # user configuration scope
  - name: "user"
    path: "~/.spack"
    optional: true
    when: '"SPACK_DISABLE_LOCAL_CONFIG" not in env'

  # site configuration scope
  - name: "site"
    path: "$spack/etc/spack/site"
    optional: true

  # system configuration scope
  - name: "system"
    path: "/etc/spack"
    optional: true
    when: '"SPACK_DISABLE_LOCAL_CONFIG" not in env'
```

## Seeing and overriding scopes

As before, you can see precedence with `spack config scopes`

```
> spack config scopes -vp
command_line     internal      active
spack            path          active  /Users/gamblin2/Workspace/src/spack/etc/spack/
user             include,path  active  /Users/gamblin2/.spack/
site             include,path  active  /Users/gamblin2/Workspace/src/spack/etc/spack/site/
defaults         path          active  /Users/gamblin2/Workspace/src/spack/etc/spack/defaults/
defaults:darwin  include,path  active  /Users/gamblin2/Workspace/src/spack/etc/spack/defaults/darwin/
defaults:base    include,path  active  /Users/gamblin2/Workspace/src/spack/etc/spack/defaults/base/
_builtin         internal      active
```

`system` doesn't appear here b/c it's optional and it will only appear when it exists on the host machine. 

There is a new `-v` / `--verbose` option to `spack config scopes` that will show the scope type and whether a scope is active (i.e. not overridden). If a scope is overridden, it still appears in the output, but it is dimmed, e.g.:

<img width="1129" height="360" alt="Screenshot 2025-10-26 at 2 33 07 AM" src="https://github.com/user-attachments/assets/4fde1b67-8a3d-4df1-83c2-f2cd1e3026e6" />

or simply:

<img width="1069" height="210" alt="Screenshot 2025-10-26 at 2 33 33 AM" src="https://github.com/user-attachments/assets/a8bcc72b-f416-4ba4-92ea-e1aa2574447a" />

You can see that the `include::` in the environment here overrides everything but defaults. This allows users to make insulated environments that don't use any local configuration. The fact that defaults remain is a special case, as `::` in other contexts *can* override things set in defaults.  Spack can't function without the `defaults` configuration scope, so here, we disallow it. You can still override defaults with `::` in other config sections, you just have to be more specific.

## `site` and `spack` scopes

What was the site scope (`$spack/etc/spack`) is now the `spack` scope, and the `site` scope is at (`$spack/etc/spack/site`) so that scripts using, e.g. `spack config --scope site set ...` will still work. The `spack` scope is *higher* precedence than the `user` scope, something people have wanted for a long time, while the `site` scope is lower precedence than `user`, as it always has been.



- [x] allow scopes to specify their own name in config
- [x] rework incremental config initialization to work when configs are included recursively
- [x] remove hard-coded scope initialization
- [x] rework defaults scope so that precedence is preserved (already had `base` -- just move everything there to get this)
- [x] tests
- [x] Add ways to override scopes easily
- [x] Think about what `::` means and whether we should preserve defaults (currently we do)

tgamblin committed:
config scopes can specify their own name

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
rework existing scopes as config

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
config: Make spack instance scope top priority

- [x] implement scope overrides
- [x] all scopes flow from $spack/etc/spack
- [x] site scope is below user scope and above system scope
- [x] envs and config with `spack:include:[]` can override user/site/system
- [x] `include::` is handled in config merge by reomoving overridden scopes

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

haampie commented on etc/spack/defaults/include.yaml:
```diff
@@ -1,4 +1,29 @@
+  - name: "user"
+    path: "~/.spack"
+    optional: true
+    when: '"SPACK_DISABLE_LOCAL_CONFIG" not in env'
```
I don't like that `eval` becomes part of default Spack. Can't it be `when: environment: ...` or something?

tgamblin replied:
The only new thing here is `name:`. That eval/conditional `when` construct has been in spack for `definitions:` since around when we introduced environments: https://spack-tutorial.readthedocs.io/en/latest/tutorial_stacks.html#conditional-definitions. We just reused it for `include:`

Or do you mean that default config shouldn’t use it? I suppose we could add something special, but I think it’s pretty simple as is… do you want something like this?

```
when_set: ENV_VAR
```

haampie replied:
Yeah, I just think it makes Spack look bad if it *needs* `eval` to work at all. On develop it's a bit of a hidden feature, so I don't mind and don't use it.

tgamblin replied:
I tried this out and I feel like `when_set` (or really any special case config for this) makes the code more complicated than it's worth when it's pretty simple to write:

```yaml
    when: '"SPACK_DISABLE_LOCAL_CONFIG" not in env'
```

It's pretty readable. `when_set` only shortens the line to:

```yaml
    when_set: SPACK_DISABLE_LOCAL_CONFIG
```

and it complicates the already somewhat complicated include logic to have two different condition types.  With `when_set` at least you have to think about "what if they supply `when:` and `when_set:`?".. and adding `environment:` as an optional thing below when makes the typing of `when` complicated. 

I'm not sure what's fundamentally *bad* about eval here... I think we can add something else later if you really want this to disappear from default config.

alalazo commented:
> All configuration stems from $spack/etc/spack/defaults/include.yaml, which looks like this:

I have a few questions on this:

1. With respect to https://github.com/spack/spack/pull/47615#issuecomment-3159967238 this doesn't take into consideration the third request, which is to leave custom user configuration out of the Spack installation directory. Is that not a concern anymore? 
2. If "all configuration stems from `$spack/etc/spack/defaults/include.yaml`" does it mean that `includes.yaml` is a special configuration file that exists only in that scope, and will not be merged with `includes.yaml` in other scopes?
    * If yes, we are special casing this configuration file, which might be confusing. At that point it's the same as the file I proposed in the comment, but it's within the Spack installation by default.
    * If no, we might merge include files that exclude themselves as scopes. How do we deal with that?

alalazo commented:
See for instance this in terms of semantics. I cannot explain this behavior:
```console
$ cat ~/.spack/include.yaml
include::
  - name: "site"
    path: "/tmp"
$ spack config scopes -p -t include
user             /home/culpo/.spack/
site             /tmp/
_defaults:linux  /home/culpo/PycharmProjects/spack/etc/spack/defaults/linux/
_defaults:base   /home/culpo/PycharmProjects/spack/etc/spack/defaults/base/
$ spack config blame include
---                                                                   include:
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:8   - name: user
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:9     path: ~/.spack
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:10    optional: True
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:11    when: '"SPACK_DISABLE_LOCAL_CONFIG" not in env'
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:14  - name: site
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:15    path: $spack/etc/spack
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:16    optional: True
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:19  - name: system
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:20    path: /etc/spack
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:21    optional: True
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:22    when: '"SPACK_DISABLE_LOCAL_CONFIG" not in env'
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:25  - path: ${platform}
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:26    optional: True
/home/culpo/PycharmProjects/spack/etc/spack/defaults/include.yaml:29  - path: base
/home/culpo/.spack/include.yaml:2                                     - name: site
/home/culpo/.spack/include.yaml:3                                       path: /tmp
```
Problems I see:
1. I cannot override the scopes from default?
2. If I could, it's not clear how a scope that exclude itself, or previous scopes, is merged
3. `spack config blame` makes it look like `site` is appended to the list of scopes, `spack config scopes` gives a different precedence
4. Even more fun if we name a scope with something already parsed, like "user"

To be clear, I'm not tied to the solution I proposed (happy to scrap that), but I still don't see how we can define scopes in a config that is merged across the scopes it's being defining.

psakievich commented:
https://github.com/spack/spack/pull/51165 is a related effort.

tgamblin committed:
`config scopes`: show scope types and active/overridden status

- [x] Add a new option, `-v` to `spack config scopes` that shows scope types
      and whether the scope has been overridden
- [x] Show overridden scopes as gray in `spack config scopes` output

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
fix spack config scope types test

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
add test for include overrides

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on etc/spack/README.md:
```diff
@@ -0,0 +1,6 @@
+# Spack main config scope
+
+This Spack's top-level configuration scope. Everything except Spack's bare-bones
+default confiugration is included from here.
```
```suggestion
default configuration is included from here.
```

alalazo commented on lib/spack/spack/cmd/config.py:
```diff
@@ -90,6 +91,14 @@ def setup_parser(subparser: argparse.ArgumentParser) -> None:
         choices=("all", "env", "include", "internal", "path"),
         help="list only scopes of the specified type(s)\n\noptions: %(choices)s",
     )
+    scopes_parser.add_argument(
```
(nitpick) Maybe a header on the first line would be nice for readability:
```console
% spack config scopes -vv
command_line     internal      active   
spack            path          active   
user             include,path  active   
site             include,path  active   
defaults         path          active   
defaults:darwin  include,path  active   
defaults:base    include,path  active   
_builtin         internal      active   
```

tgamblin replied:
Done

alalazo replied:
Seems like commits were not pushed upstream?

tgamblin replied:
it's there -- the headers only appear when there is more than one column.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -1342,6 +1412,9 @@ def get(path: str, default: Optional[Any] = None, scope: Optional[str] = None) -
     return CONFIG.get(path, default, scope)
 
 
+_set = set  #: save this before defining set -- maybe config.set was ill-advised :)
```
(Not in this PR) Yeah, this is bad. If this function is not in the package API, we might want to deprecate and inline it, together with `spack.config.get`

tgamblin replied:
We could, though it is used all over the codebase.

psakievich commented on lib/spack/spack/config.py:
```diff
@@ -494,9 +517,30 @@ def push_scope(
+            priority: priority of the scope
+
+        """
+        for _ in self.push_scope_incremental(scope=scope, priority=priority, _depth=_depth):
+            pass
```
I stared at this for a while. I'm assuming this is updating the singleton config scope due to the mutator decorator?

tgamblin replied:
We have parts of config (e.g., remote config fetches) that need other parts of config. The least bad way I thought of to deal with this is to allow the configuration to use itself in a partially initialized state. That is possible when config is just a global variable, but since it is a lazy singleton, it would appear empty until *after* all config initialization ran.

The solution was to make the singleton accept not just a single initialization function, but a generator, and the value yielded by the generator is repeatedly assigned to the singleton's `_instance` during initialization, which means that the singleton can be used by parts of init. You can see the pattern in https://github.com/spack/spack/blob/8f48906/lib/spack/spack/llnl/util/lang.py#L724-L736;  `push_scope_incremental` is the analog for `push_scope` -- `push_scope` just calls it and consumes the iterator since it doesn't need to update the `_instance` variable anymore.

A possibly simpler way to do this would just be to have the singleton create the instance *then* set it up, but passing a generator instead of a function kind of more general, in that the user *can* just pass a single function, or they can break their process up into steps, deepening on what they want. If you separate the two (create and setup) then the user always has to pass two things, which is less ergonomic IMO.

psakievich replied:
That makes sense to me. I think it would be helpful to document some of this explanation in the doc string or in a comment inside the function.

tgamblin replied:
Tried to beef this up in at the callsite in `push_scope` and in the `push_scope_incremental` docstring

psakievich reviewed (dismissed):
Playing around with this today. Working well on all my tests so far. I tested 

- config isolation in an environment `spack:include::[]` 
- how that behaves with `spack -E`
- new option `spack config scopes -v`

One thing I noticed is running `spack config scopes -v` does not issue a warning with duplicate scopes of the same name and just ommitts the duplicate scope. Other commands do issue a warning like this

``` console
$ spack cd -e
==> Warning: Ignoring duplicate included scope: test
```

 I think we should unify that behavior or indicate in some manner that a scope is getting skipped. So far this is the only issue I've found. I'll do some more testing, but so far this is looking really good.
 
 NVM I missed it. The warning DOES show up. I just didn't see it.

psakievich commented on lib/spack/spack/config.py:
```diff
@@ -494,9 +517,30 @@ def push_scope(
+        """Add a scope to the Configuration, at a given priority.
+
+        This version of push_scope yields included scopes incrementally, so that their
+        data can be used by higher priority scopes during config initialization.
```
This is duplicated between the two functions. Assuming it is only correct for `push_scope_incremental` and not this function?

tgamblin replied:
fixed

psakievich commented:
> If yes, we are special casing this configuration file, which might be confusing. At that point it's the same as the file I proposed in the comment, but it's within the Spack installation by default.

@alalazo I think not being in the Spack installation was the only thing I didn't like about your proposal. To me keeping the special case internal to the Spack installation helps guard the feature I really want, which is sharing is optional, dependency on the home directory is optional.

psakievich reviewed (commented):
I don't see anything beyond docs truly missing from this PR.  I would really like it if there was a way to pop a scope from the CLI.  

For instances, I would love to be able to download a spack instance, `spack config scopes pop user` and just have it get commented out or something. Explicitly disable it for the spack instance. Personally I would love an easy way to disable/pop everything but `site` and the `spack` scopes out the gate.

This maybe better suited for the `spack isolate` command we discussed though.

alalazo commented:
> I think not being in the Spack installation was the only thing I didn't like about your proposal. To me keeping the special case internal to the Spack installation helps guard the feature I really want, which is sharing is optional, dependency on the home directory is optional.

Agreed. The comment you're replying to is from August, and is outdated. We agreed on the current implementation in the last two-three meetings about this PR :slightly_smiling_face:

alalazo commented on lib/spack/spack/llnl/util/tty/colify.py:
```diff
@@ -63,11 +64,21 @@ def config_variable_cols(elts, console_width, padding, cols=0):
+        # specified, otherwise, use the requested columns and overflow.
         config = configs[0]
-
+        if cols:
```
Am I correct that changes to this file are a separate bugfix for forcing the number of columns? In case:
- Does it have unit tests?
- Is it worth extracting it as a standalone (to be backported)?

tgamblin replied:
Basically:
* This works fine for a fixed column width of 1
* Requested column width is zero for (nearly?) all calls to colify
* `colify_table` is the only place (in spack at least) where this matters, because it fixes the column count, and only really uses `colify` to get its variable-width behavior.

By default the algorithm searches for a valid column configuration from 1 to some max number of columns, and if no configuration is valid, it doesn't matter -- the first element here is a single column list, which is fine.

If you fix the column count and run this and there is no valid configuration, picking the first element isn't correct -- so now we just go back and re-flow the table with the requested number of columns.  We *could* skip the search altogether in this case, I suppose. It would cut out unnecessary work (since what is displayed will be the same).

Probably it should have tests. I don't think it's worth a back port but it can be standalone.

tgamblin replied:
Factored it out and added tests here: https://github.com/spack/spack/pull/51488

alalazo replied:
Thanks!

alalazo reviewed (changes requested):
The PR LGTM 👍 I think the choice to always keep the default configuration is the correct one.

The main request is to add docs for the new behavior (can be minimal, and expanded in a following PR, but we should avoid forgetting it).

A minor comment is that I find the difference between:
```console
$ spack config scopes -v
```
and
```console
$ spack config scopes -v packages
```
confusing. I expected the second command to grey out anything that doesn't affect packages configuration, but it seems it only adds a `packages.yaml` to some of the  paths. Right now it feels a bit redundant, and similar to `-p`. Would it be easy to make the extra "packages" grey out more scopes if they don't matter for that section of the configuration?

tgamblin committed:
typo

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`config scopes`: Add column headers and `absent` status

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
add additional tests

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -677,11 +749,16 @@ def _get_config_memoized(
             # read potentially cached data from the scope.
             data = config_scope.get_section(section)
+            if data and section == "include":
+                data["include"] = data.pop("include")  # strip override
 
```
@copilot In Spack the string keys within the dictionary have an `override` attribute that indicates the original yaml contained `include::` and not `include:`.  This isn't a no-op -- it actually changes the instance of the string object used as a key in the dictionary to a plain string instead of one with the attribute.

tgamblin committed:
fix copy/paste error in docstring

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
fix test on windows

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
allow path_override_env_var to preserve SPACK_USER_CONFIG_PATH

This was lost in the restructure; allow it as an optional override var on includes.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin commented on lib/spack/spack/cmd/config.py:
```diff
@@ -261,47 +270,77 @@ def config_list(args):
+                if section_path and os.path.exists(section_path)
+                else f"{scope.path}{os.sep}"
+            )
+        result.append(section_path or " ")
```
fixed

tgamblin committed:
config scopes in config: add docs and comments

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on lib/spack/docs/configuration.rst:
```diff
@@ -66,13 +66,14 @@ From lowest to highest precedence:
+   The systm scope overrides the defaults scope.
+   It can be used for settings idiosyncratic to a particular machine, such as the locations of compilers or external packages.
+   Be careful when modifying this scope, as changes here affect all Spack users on a machine.
+   Before putting configuration here, instaed consider using the ``site`` scope, which only affects the spack instance it's part of.
```
```suggestion
   Before putting configuration here, instead consider using the ``site`` scope, which only affects the spack instance it's part of.
```

alalazo commented on etc/spack/include.yaml:
```diff
@@ -0,0 +1,18 @@
+include:
+  # user configuration scope
+  - name: "user"
+    path_override_env_var: SPACK_USER_CONFIG_PATH
```
(question, feel free to close if that seems ok as is) Should we go for a shorter name? For instance `path_env_var`?

tgamblin replied:
I think we should stick with a clear one — I like it that this says it’s about the path and that it’s an override.

alalazo commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -92,3 +80,163 @@ We would then configure the ``include.yaml`` file as follows:
+
+.. warning::
+
+   Using ``name:` to override the ``defaults`` scope can have *very* unexpected consequences and is not advised.
```
```suggestion
   Using ``name:`` to override the ``defaults`` scope can have *very* unexpected consequences and is not advised.
```

alalazo commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -81,16 +81,37 @@ def test_config_scopes(path, types, mutable_mock_env_path):
+    lines = output.split("\n")
+
+    print(lines)
+
```
```suggestion
```

alalazo commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -81,16 +81,37 @@ def test_config_scopes(path, types, mutable_mock_env_path):
-    output = config(*scopes_cmd).split()
-    assert not output or all(":" in x for x in output)
+@pytest.mark.parametrize("type", ["path", "include", "internal", "env"])
+def test_config_scopes_include(type):
```
A one liner docstring explaining what is tested might help our future selves 🙂

tgamblin replied:
fixed.

alalazo commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -81,16 +81,37 @@ def test_config_scopes(path, types, mutable_mock_env_path):
-    assert "_builtin" not in output
-    assert "site" not in output
+    lines_by_scope_name = {line.split()[0]: line for line in lines}
+    print(lines_by_scope_name)
```
```suggestion

```

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1223,27 +1224,144 @@ def test_default_install_tree(monkeypatch, default_config):
+    yield make_config
+
+
+def test_local_config_can_be_disabled(working_env, include_config_factory):
```
Same  as https://github.com/spack/spack/pull/51162#discussion_r2485833351 in this file

tgamblin replied:
done.

alalazo reviewed (commented):
Nitpick comments + a few typos. I really like:

<img width="1652" height="300" alt="Image" src="https://github.com/user-attachments/assets/1bd0f90a-99cd-40a4-8d66-7d8ead655b62" />

I think this is ready to be merged, once we fix the typos.

tgamblin committed:
Updates based on PR review

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>
